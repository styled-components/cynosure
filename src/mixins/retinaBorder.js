// @flow

type Direction = 'top' | 'right' | 'bottom' | 'left'

interface BorderStyles {
  content: '""';
  position: 'absolute';
  top?: number | string;
  right?: number | string;
  bottom?: number | string;
  left?: number | string;
  width?: number | string;
  height?: number | string;
  color?: string;
  border?: string;
  borderTop?: string;
  borderRight?: string;
  borderBottom?: string;
  borderLeft?: string;
  transformOrigin?: string;
  transform?: string;
  boxSizing?: 'border-box';
}

/**
 * 1px border for Retina screen.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   position: 'relative',
 *   '&::after': {
 *     ...retinaBorder(),
 *   },
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   position: relative;
 *   &::after {
 *     ${retinaBorder()}
 *   }
 * `
 *
 * // CSS as JS Output
 *
 * div: {
 *   position: 'relative',
 *   '&::after': {
 *     content: '""',
 *     position: 'absolute',
 *     top: 0,
 *     left: 0,
 *     width: '200%',
 *     height: '200%',
 *     border: '1px solid #c7c7c7',
 *     transformOrigin: '0 0',
 *     transform: 'scale(0.5)',
 *     boxSizing: 'border-box',
 *   }
 * }
 */

function retinaBorder(
  color?: string = '#c7c7c7',
  direction?: Direction,
): Object {
  const styles: BorderStyles = {
    content: '""',
    position: 'absolute',
  }

  switch (direction) {
    case 'top':
      Object.assign(styles, {
        top: 0,
        right: 0,
        left: 0,
        height: '1px',
        borderTop: `1px solid ${color}`,
        transformOrigin: '0 0',
        transform: 'scaleY(0.5)',
      })
      break
    case 'right':
      Object.assign(styles, {
        top: 0,
        right: 0,
        bottom: 0,
        width: '1px',
        borderRight: `1px solid ${color}`,
        transformOrigin: '100% 0',
        transform: 'scaleX(0.5)',
      })
      break
    case 'bottom':
      Object.assign(styles, {
        right: 0,
        bottom: 0,
        left: 0,
        height: '1px',
        borderBottom: `1px solid ${color}`,
        transformOrigin: '0 100%',
        transform: 'scaleY(0.5)',
      })
      break
    case 'left':
      Object.assign(styles, {
        top: 0,
        bottom: 0,
        left: 0,
        width: '1px',
        borderLeft: `1px solid ${color}`,
        transformOrigin: '0 0',
        transform: 'scaleX(0.5)',
      })
      break
    default:
      Object.assign(styles, {
        top: 0,
        left: 0,
        width: '200%',
        height: '200%',
        border: `1px solid ${color}`,
        transformOrigin: '0 0',
        transform: 'scale(0.5)',
        boxSizing: 'border-box',
      })
      break
  }

  return styles
}

export default retinaBorder
