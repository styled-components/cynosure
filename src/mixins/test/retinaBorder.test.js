// @flow
import retinaBorder from '../retinaBorder'

describe('retinaBorder', () => {
  it('should be the default when no color and direction are passed', () => {
    expect(retinaBorder()).toMatchSnapshot()
  })

  it('should set the top border', () => {
    expect(retinaBorder('#c7c7c7', 'top')).toMatchSnapshot()
  })

  it('should set the right border', () => {
    expect(retinaBorder('#c7c7c7', 'right')).toMatchSnapshot()
  })

  it('should set the bottom border', () => {
    expect(retinaBorder('#c7c7c7', 'bottom')).toMatchSnapshot()
  })

  it('should set the left border', () => {
    expect(retinaBorder('#c7c7c7', 'left')).toMatchSnapshot()
  })
})
