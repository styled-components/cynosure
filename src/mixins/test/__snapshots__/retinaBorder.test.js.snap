// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retinaBorder should be the default when no color and direction are passed 1`] = `
Object {
  "border": "1px solid #c7c7c7",
  "boxSizing": "border-box",
  "content": "\\"\\"",
  "height": "200%",
  "left": 0,
  "position": "absolute",
  "top": 0,
  "transform": "scale(0.5)",
  "transformOrigin": "0 0",
  "width": "200%",
}
`;

exports[`retinaBorder should set the bottom border 1`] = `
Object {
  "borderBottom": "1px solid #c7c7c7",
  "bottom": 0,
  "content": "\\"\\"",
  "height": "1px",
  "left": 0,
  "position": "absolute",
  "right": 0,
  "transform": "scaleY(0.5)",
  "transformOrigin": "0 100%",
}
`;

exports[`retinaBorder should set the left border 1`] = `
Object {
  "borderLeft": "1px solid #c7c7c7",
  "bottom": 0,
  "content": "\\"\\"",
  "left": 0,
  "position": "absolute",
  "top": 0,
  "transform": "scaleX(0.5)",
  "transformOrigin": "0 0",
  "width": "1px",
}
`;

exports[`retinaBorder should set the right border 1`] = `
Object {
  "borderRight": "1px solid #c7c7c7",
  "bottom": 0,
  "content": "\\"\\"",
  "position": "absolute",
  "right": 0,
  "top": 0,
  "transform": "scaleX(0.5)",
  "transformOrigin": "100% 0",
  "width": "1px",
}
`;

exports[`retinaBorder should set the top border 1`] = `
Object {
  "borderTop": "1px solid #c7c7c7",
  "content": "\\"\\"",
  "height": "1px",
  "left": 0,
  "position": "absolute",
  "right": 0,
  "top": 0,
  "transform": "scaleY(0.5)",
  "transformOrigin": "0 0",
}
`;
